### PR TITLE
Mark abstract classes as such 

### PR DESCRIPTION
**Pull request by thk123**
When parsing Java classes, mark them as abstract if they are either abstract or an interface. This will be used when replacing abstract classes and interfaces with concrete implementations. 

Added unit tests that check the `typet` associated with the `symbolt` for different classes. 

I won't advance the test-gen-support pointer until the whole issue is complete, but will as suggested by @smowton, have a PR that proves that they work so they don't block other developments.

### CODE REVIEW

**Comment by peterschrammel** [comment]
@thk123, please rebase to make CI happy.

File: src/util/irep_ids.def [hunk 1]
```diff
@@ -816,6 +816,7 @@ IREP_ID_ONE(basic_block_covered_lines)
 IREP_ID_ONE(is_nondet_nullable)
 IREP_ID_ONE(array_replace)
 IREP_ID_ONE(java_array_access)
+IREP_ID_ONE(is_abstract_class)
```
Comment (mgudemann) [inline src/util/irep_ids.def]: is this needed? there is `ID_abstract` which is used in the Java front-end
```
  // Construct a methodt, which lives within the class type; this object is
  // never used for anything useful and could be removed
  class_typet::methodt method;
  method.set_base_name(m.base_name);
  method.set_name(method_identifier);
  method.set(ID_abstract, m.is_abstract);
  method.set(ID_is_virtual, is_virtual);
  method.type()=member_type;
  if(is_constructor(method))
    method.set(ID_constructor, true);
```

File: unit/Makefile [hunk 2]
```diff
@@ -14,7 +15,9 @@ include ../src/common
 cprover.dir:
 	$(MAKE) $(MAKEARGS) -C ../src
 
-LIBS += ../src/ansi-c/ansi-c$(LIBEXT) \
+LIBS += ../src/java_bytecode/java_bytecode$(LIBEXT) \
+        ../src/miniz/miniz$(OBJEXT) \
+        ../src/ansi-c/ansi-c$(LIBEXT) \
```
Comment (mgudemann) [inline unit/Makefile]: please sort alphabetically

Comment (thk123) [inline unit/Makefile]: Sadly not possible - the order effects whether it compiles(!) - https://stackoverflow.com/questions/45135/why-does-the-order-in-which-libraries-are-linked-sometimes-cause-errors-in-gcc

Comment (mgudemann) [inline unit/Makefile]: ok

**Comment by thk123** [comment]
@mgudemann done - ready for re-review

**Comment by thk123** [comment]
Corresponding test-gen PR: diffblue/test-gen#699

**Comment by thk123** [comment]
Going to wait for the master->TGS PR (#1094) to go in first just as it hard to get in!

**Pull request merged**